docs/proposals/vip-3.md

```diff
@@ -0,0 +1,246 @@
+   flexibility, since the client would not remain stuck with whatever resource name was returned by the cluster 
+   discovery call done at startup time (and periodically thereafter), but rather could adapt rapidly.
+
+These options provide different levels of granularity of control and speed of change, with finer granularity and faster
+change carrying greater complexity cost.
+
+### Auth/Auth
+
+The Rust server will open a port and receive connections from clients, so it will need to authenticate and authorize
+them. Currently, the auth/auth logic is all in Java, so there needs to be a solution to doing it on the Rust side. There
+are two main alternatives:
+
+1. Design an abstraction which can be implemented by a custom Rust module, so that the Rust server can interrogate an
+   API and find out if a given principal/token/etc is allowed to access a given resource.
+2. Implement a mechanism by which the Rust server can call into the Java server, ask whether a given principal/token/etc
+   is allowed to access a given resource, let the Java server answer that with the current abstraction that exists over
+   there, return that response to Rust, and finally let Rust cache this decision so that known principals need not do
+   this again. In this scheme, there needs to also be a mechanism to periodically refresh the authorization decision
+   (e.g. once a minute), in case it may have changed.
+
+Given that Venice already requires the operator to provide a Java implementation of the authorization implementation, it
+is probably easier to leverage this one and only implementation in Rust as well (via approach 2) rather than require the
+operator to implement their authorization in two places. Furthermore, approach 2 has the potential to be more efficient,
+since the authorization logic need not run within the hot path of every single request. For these reasons, approach 2 is
+preferred.
+
+## Development Milestones
+
+The milestones for the MVP would be the following:
+
+1. Build a simple RocksDB wrapper which gets packaged into a jar and exposes the same JNI APIs as RocksJava. This can be
+   built with a single architecture at first. Print "hello world" when JNI is called successfully.
+2. Swap RocksJava for the wrapper inside Venice. We can do a micro-benchmark at this stage to validate the assumption 
+   that JNI via Rust is no worse than JNI via C++.
+3. Integrate the chosen crate for adding gRPC support, and do a "hello world" of the RPC.
+4. Add server configs to enable the Rust gRPC module (including basic service disco announcement).
+5. Implement the batch get support.
+6. Implement support for auth/auth.
+7. Multi-architecture (Linux + Mac) support.
+8. (Possibly optional) implement quota.
+9. (Possibly optional) more advanced/granular service disco control.
+
+Beyond the MVP, the following scope expansions are natural followups:
```

Comment:
If I understand this idea correctly, this should achieve:
(1) full ingestion isolation without hybrid store continuous ingestion concern.
(2) Simplify write path and state transition (back to non-II state without maintaining state machine between main and forked)
The concern might mostly be:
(1) Effort to develop the Rust impl (Well we are all young so should be easy to learn, but for site maintenance perspective I am not sure if SRE will have concern on this. My understanding is if keep providing enough metrics/loggings it should be fine).
(2) I know other ppl chat about C++ vs Rust about pure performance, I am not super concern about that...as we were fine with Java today. The one minor concern is https://github.com/rust-rocksdb/rust-rocksdb I think it would be the wrapper of RocksDB to use in Rust Read Process. It is not maintained by RocksDB main contributor, but rather a community support, as it is yet another wrapper on top of original C++ implementation. I am not sure if the support is up to date and there is any issue / future issue with that. 

Also, since we have limited dev resources today, if we are determined to prioritize this in either R/W team, we should pivot our plan from exploring II to this. (Which is partially done, but not yet)

Reply:
Although this project is orthogonal to blob ingestion idea, which is great. 
I am seeing stronger need to have blob ingestion in our Venice support -> for both Da Vinci and Server

Reply:
You bring up several important points here, so let me try to address each of them...

Regarding II, the way I see it, if we choose to go forward with the Rust server proposal, then we should cancel the development of hybrid II. As for batch II, we need to make a judgment call... batch II has proven useful in some of our workloads, and if it is close to being stable, then it would be great to leverage it for the time being, since the Rust server may take 6-12 months to land in production. If batch II is far from being stable and we need another 6 months to stabilize it, then of course it's not as attractive... we don't know that for sure, but we'll need to make a judgment call. Besides, even if we made the call that batch II in servers is too troublesome to stabilize and we want to abandon it, we are still using batch II for DVC, and that still needs to be supported going forward (unless/until blob ingestion comes around, I guess). In the end state, when the Rust server is available, at that point it seems logical that we may want to stop using batch II for server, but I'm not 100% sure... there may be a reason to keep it around (e.g. batch vs hybrid isolation perhaps).

I am hoping to be able to use the crate coming of the rust-rocksdb repo you linked. It seems to have gotten a fair amount of development, with the latest commit being not that long ago, so it looks promising. If something is missing or unsatisfactory, we can consider contributing to it, or other alternatives...